18273+18558 Extension Request Cleanup and Tests

leodube-aot opened the pull request:
*Issue #:* /bcgov/entity#18273 and /bcgov/entity#18558

*Description of changes:*
- Connected `alreadyExtended` and `requestExpired` for `AgmExtensionEvaluation.vue`.
- Implemented `extensionRequestValid` and emit result to parent component.
- Added `evaluateResult` flag to only show extension request warning box and duration of extension text field on valid extension request.
- Moved scenario logic into separate functions called from a `checkEligibility()` function.
- Added unit tests for `ExtensionRequest` component to cover every scenario.
- Updated unit tests for `AgmExtensionEvaluation`
- Created unit tests for `date-utilities`


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

leodube-aot commented on src/components/AgmExtension/AgmExtensionEvaluation.vue:
```diff
@@ -10,7 +10,7 @@
     <template #content>
       <v-card
-        v-if="!data.isEligible"
+        v-if="evaluateResult && !data.isEligible"
```
Added `evaluateResult` to only display the yellow message box, as well as the custom text in "Duration of Extension" when the extension request is valid (i.e. all fields for a specific scenario are entered)

severinbeauvais replied:
This partly resolves #18558. Want to take that ticket and fix it here?

severinbeauvais replied:
PS You don't have to.

leodube-aot replied:
I don't mind fixing it while I'm already working in the code!

leodube-aot commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -180,27 +183,18 @@ import { DateUtilities } from '@/services'
-
-  @Emit('update:data')
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  private emitData (data: AgmExtEvalIF): void {}
```
Moved emit function to bottom of file

leodube-aot commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -209,14 +203,36 @@ export default class ExtensionRequest extends Vue {
+  get extensionRequestValid (): boolean {
+    const isFirstAgmValid = this.data.isFirstAgm && !!this.data.agmYear
+    const isSubsequentAgmValid = (
+      this.data.isFirstAgm === false &&
```
I specifically check `=== false` a few times in this file. That's because some values in the `data` object can be `null`, but that doesn't mean they are valid. 

For example, the answers to "Is this the first AGM?" can be "Yes" or "No", with values `true` and `false` respectively. To check `isSubsequentAgmValid`, then the "No" field must be ticked (which evaluates to false). However when neither field is ticked (not a valid extension request), it remains `null`. Therefore there can be no ambiguity between null and false.

<img width="608" alt="Screenshot 2023-11-17 at 1 22 49 PM" src="https://github.com/bcgov/business-filings-ui/assets/122323255/3398ed30-5bc2-43c9-b4ca-5921595176b8">

leodube-aot commented on tests/unit/date-utilities.spec.ts:
```diff
@@ -0,0 +1,190 @@
+import { DateUtilities } from '@/services'
```
I created this test file for the `date-utilities`, then copied over the tests from `date-mixins.spec.ts` and modified them a bit to get them to work. The `date-mixins` are deprecated but are still used in a few places in the application.

At the bottom I added new tests for `addMonthsToDate` and `subtractDates`

leodube-aot commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-585-c2c631r7.web.app

severinbeauvais commented:
This should probably throw an error, I think?

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/d216a304-2e69-4f64-9161-ad4c4a088c74)

leodube-aot commented:
@severinbeauvais Yeah you're correct, there should be more clear error handling. I'll get that fixed. The one field I can confirm we don't validate is the Intended AGM date (based on what Mihai has said), but perhaps at should still not allow dates in the past for that one.

severinbeauvais commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -255,149 +271,206 @@ export default class ExtensionRequest extends Vue {
-  onIsPrevExtensionChanged (): void {
-    /* eslint-disable brace-style */
+  /** Eligibility for first agm with no extension requested this year */
+  /* eslint-disable brace-style */
```
What's wrong with the brace style, and should you re-enable this check below this  method?

leodube-aot replied:
Mainly the linter expects the next conditional block to start right away:

```
if () {
} else {
}
```

Which doesn't really allow for comments between them like this:

```
// if block
if () {
}

// else block
else {
}
``` 

Since the logic is a bit tricky to understand I separated each block, but it's mainly a stylistic/readability choice. I could also have the comments inside the conditional blocks but I find it slightly less readable.

But yeah it's probably good to re-enable the brace style afterwards, I'll do that.

severinbeauvais replied:
Sounds good. You may resolve this conversation after re-enabling the eslint rule.

severinbeauvais commented on src/components/AgmExtension/AgmExtensionEvaluation.vue:
```diff
@@ -108,6 +108,7 @@ import { DateUtilities } from '@/services'
 })
 export default class AgmExtensionEvaluation extends Vue {
   @Prop({ required: true }) readonly data!: AgmExtEvalIF
+  @Prop({ default: false }) readonly evaluateResult!: boolean
```
This is specific enough, and not part of "data", that I'm OK with this being a new prop 👍

leodube-aot commented:
**Newest changes:**
- Fixed bcgov/entity#18558
- Added more validation to the text and date picker components, also made it so that if the extension request is invalid, the specific section(s) that are invalid show the red border, instead of the entire component.
- Need to fix a couple unit tests still

leodube-aot commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -11,9 +11,12 @@
       <div
         class="px-6 py-7"
-        :class="{ 'invalid-section': !extensionRequestValid && showErrors }"
+        :class="{ 'invalid-section': !firstSectionValid && showErrors }"
```
Added specific valid flags for each section of the component to highlight specific areas that are not valid.

leodube-aot commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -11,9 +11,12 @@
       >
-        <v-row no-gutters>
+        <v-row
+          v-if="!is30MonthsAfterIncorp"
```
The UXPin says that if incorporation date is 30 months ago, it is safe to assume that it is not the first agm extension, so hide the radio group and only show the `agmYear` textbox.

leodube-aot commented on src/interfaces/agm-ext-eval-interface.ts:
```diff
@@ -25,7 +25,11 @@ export const EmptyAgmExtEval: AgmExtEvalIF = {
+  prevAgmDate: null,
+  isPrevExtension: null,
+  prevExpiryDate: null,
+  intendedAgmDate: null,
```
Needed to explicitly set values for the `EmptyAgmExtEval`, otherwise the date pickers don’t update the dates properly. I think it is likely because if you `@Watch('data.intendedAgmDate')` and `data.intendedAgmDate` is undefined, when it initially gets set it doesn't seem to trigger the watcher.

severinbeauvais replied:
You are correct. If a variable is undefined, it's not reactive. It's like it doesn't exist (which is kind of the definition of undefined 😀 ).

severinbeauvais replied:
If `prevAgmDate`, `isPrevExtension`, `prevExpiryDate` and `intendedAgmDate` are no longer optional then please remove the "?"s in the interface above. Thx.

severinbeauvais replied:
LGTM

leodube-aot commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-585-c2c631r7.web.app

leodube-aot commented on src/components/AgmExtension/AgmExtensionEvaluation.vue:
```diff
@@ -19,78 +19,93 @@
-        class="px-6 pt-7"
+      <div
+        v-else-if="!evaluateResult"
+        style="overflow: auto;"
```
Yeah so the issue I was trying to resolve is that when only displaying the message box, the bottom margins get hidden in the overflow (a bit tough to see in these screenshots, but the message box is aligned at the bottom of the v-card).

<img width="868" alt="Screenshot 2023-11-20 at 10 30 05 AM" src="https://github.com/bcgov/business-filings-ui/assets/122323255/2a01edaa-0239-45c8-bf5a-eea96ad2de53">
<img width="871" alt="Screenshot 2023-11-20 at 10 30 22 AM" src="https://github.com/bcgov/business-filings-ui/assets/122323255/2c0e34b9-c404-4bc2-9790-4732ceb6fc50">

And adding padding will just expand the size of the message box. I tried `d-flex` and `flex-grow-1` but they didn't seem to work. 

I'll add a style property to base.scss and use that instead.

severinbeauvais replied:
LGTM

leodube-aot commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-585-c2c631r7.web.app

severinbeauvais commented:
Leo, does this PR complete everything for AGM Extensions? If not, what's left?

severinbeauvais commented:
Leo, should there be a check that the intended AGM date is not later than the due date? Maybe this wouldn't cause a validation error but could display the yellow box with some message. Check with Mihai and Yui.

leodube-aot commented:
@severinbeauvais I checked and there is one more formula to be added to the AGM extension calculation bcgov/entity#18579. Otherwise, this ticket has completed all the hooking up required.

For the intended date, [Mihai's comment](https://github.com/bcgov/entity/issues/18053#issuecomment-1813296398) explains why we don't validate it.

severinbeauvais commented:
> Otherwise, this ticket has completed all the hooking up required.

Yeah, it looks that way -- thank you! But I am still unable to file an AGM extension. Do we know if this is solely an API issue?

severinbeauvais commented on tests/unit/AgmExtensionEvaluation.spec.ts:
```diff
@@ -2,29 +2,35 @@ import Vue from 'vue'
         data: {
-          agmYear: 2023
-        }
+          ...EmptyAgmExtEval,
```
Nice use of destructuring 👍

severinbeauvais commented on tests/unit/ExtensionRequest.spec.ts:
```diff
@@ -2,48 +2,351 @@ import Vue from 'vue'
+
+    // Verify isPrevExtension radio group
+    expect(prevExtensionRadios.at(0).attributes('aria-checked')).toBe('false')
+    expect(prevExtensionRadios.at(1).attributes('aria-checked')).toBe('true')
```
I think there's a way to directly check if the radio button wrapper is checked but I can't think of it off the top of my head and this is OK. FYI.

severinbeauvais commented on tests/unit/ExtensionRequest.spec.ts:
```diff
@@ -2,48 +2,351 @@ import Vue from 'vue'
+    expect(wrapper.vm.data.extensionDuration).toBe(null)
+    expect(wrapper.vm.data.alreadyExtended).toBe(false)
+    expect(wrapper.vm.data.requestExpired).toBe(true)
+  })
```
Nice 👍

severinbeauvais commented on src/components/AgmExtension/ExtensionRequest.vue:
```diff
@@ -239,165 +287,230 @@ export default class ExtensionRequest extends Vue {
+  @Watch('data.isPrevExtension')
+  @Watch('data.prevExpiryDate')
+  @Watch('data.intendedAgmDate')
+  @Watch('data.currentDate')
```
Would it make sense to use `@Watch('data', { deep: true })` here, or are there data properties you don't want to watch?

leodube-aot replied:
I don't think so, since when this watcher triggers the `checkEligibility()` function, we are setting a bunch of data properties like `isElgibile`, `extensionDuration`, `requestExpired`, etc. that we don't necessarily want to watch for changes to.

PaulGarewal reviewed (approved):
LGTM, amazing job Leo 🎉

leodube-aot commented:
> Yeah, it looks that way -- thank you! But I am still unable to file an AGM extension. Do we know if this is solely an API issue?

I think it's just an API issue. The [saveFiling() function in AgmExtension](https://github.com/bcgov/business-filings-ui/blob/main/src/views/AgmExtension.vue#L336) throws an axios error: "Request failed with status code 422" when "File and Pay" is clicked, which prevents redirection to the payment page.
